src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Snippets/XmlSnippetParser.ExpansionTemplate.cs

```diff
@@ -0,0 +1,233 @@
+            var delimiterAttribute = code.Attributes().FirstOrDefault(a => a.Name.LocalName.Equals("Delimiter", StringComparison.OrdinalIgnoreCase));
+            if (delimiterAttribute != null)
+            {
+                _delimiter = delimiterAttribute.Value;
+            }
+        }
+
+        private void ReadDeclarations(XElement? declarations)
+        {
+            if (declarations == null)
+            {
+                return;
+            }
+
+            foreach (var declarationElement in declarations.Elements())
+            {
+                var editableAttribute = declarationElement.Attribute("Editable");
+                var functionElement = CodeSnippet.GetElementWithoutNamespace(declarationElement, "Function");
+                TryGetSnippetFunctionInfo(functionElement?.Value, out var functionName, out var functionParam);
+                _tokens.Add(new ExpansionField(
+                    CodeSnippet.GetElementInnerText(declarationElement, "ID"),
+                    CodeSnippet.GetElementInnerText(declarationElement, "Default") ?? " ",
+                    functionName,
+                    functionParam,
+                    editableAttribute == null || string.Equals(editableAttribute.Value, "true", StringComparison.Ordinal) || string.Equals(editableAttribute.Value, "1", StringComparison.Ordinal)));
+            }
+        }
+
+        internal ParsedXmlSnippet Parse()
+        {
+            int iTokenLen;
+            var currentCharIndex = 0;
+            var currentTokenCharIndex = 0;
+            var sps = SnippetParseState.Code;
+
+            // Associate the field id to the index of the field in the snippet.
+            var fieldNameToSnippetIndex = new Dictionary<string, int>();
+            var currentTabStopIndex = 1;
+
+            using var snippetParts = new PooledArrayBuilder<SnippetPart>();
+
+            // Mechanically ported from env/msenv/textmgr/ExpansionTemplate.cpp
+            while (currentCharIndex < _code!.Length)
+            {
+                iTokenLen = currentCharIndex - currentTokenCharIndex;
+
+                switch (sps)
+                {
+                    case SnippetParseState.Code:
+                        if (string.Equals(_code[currentCharIndex].ToString(CultureInfo.CurrentCulture), _delimiter, StringComparison.Ordinal))
+                        {
+                            // we just hit a $, denoting a literal
+                            sps = SnippetParseState.Literal;
+
+                            // copy anything from the previous token into our string
+                            if (currentCharIndex > currentTokenCharIndex)
+                            {
+                                // append the token into our buffer
+                                var token = _code.Substring(currentTokenCharIndex, iTokenLen);
+                                snippetParts.Add(new SnippetStringPart(token));
```

Comment:
Consider holding onto a `ReadOnlyMemory<char>` that references the original `_code` rathe that allocating a new string.